src/transformers/quantizers/quantizer_compressed_tensors.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from ..utils import is_torch_available, logging
+from ..utils.quantization_config import QuantizationConfigMixin
+from .base import HfQuantizer
+
+
+if is_torch_available():
+    import torch
+
+logger = logging.get_logger(__name__)
+
+
+class CompressedTensorsHfQuantizer(HfQuantizer):
+    """
+    Quantizer for the compressed_tensors package.  Loads and restores models to
+    quantized state with compressed_tensors
+    """
+
+    requires_calibration = False
+    # requires_parameters_quantization = True
+    required_packages = ["compressed_tensors"]
+
+    def __init__(self, quantization_config: QuantizationConfigMixin, **kwargs):
+        super().__init__(quantization_config, **kwargs)
+
+        from compressed_tensors.compressors import ModelCompressor
+
+        self.compressor = ModelCompressor.from_compression_config(quantization_config)
+
+    def validate_environment(self, *args, **kwargs):
+        # check torch and compressed_tensors are available, let ImportError raise otherwise
+        import torch  # noqa
+        from compressed_tensors.compressors import ModelCompressor  # noqa
+
+    def update_torch_dtype(self, torch_dtype: "torch.dtype") -> "torch.dtype":
+        if torch_dtype is None:
+            torch_dtype = torch.float16
+        elif torch_dtype != torch.float16:
+            logger.info(
+                "We suggest you to set `torch_dtype=torch.float16` for better efficiency with compressed_tensors."
+            )
+        return torch_dtype
+
+    def _process_model_before_weight_loading(self, model, **kwargs):
+        if self.quantization_config.quantization_config is not None:
+            from compressed_tensors.quantization import apply_quantization_config
+
+            apply_quantization_config(model, self.quantization_config.quantization_config)
+
+    def _process_model_after_weight_loading(self, model, resolved_archive_file, **kwargs):
+        self.compressor.decompress(model_path=resolved_archive_file, model=model)
+
```

Comment:
I have a few points I would like to highlight: 
- It would be great to load the quantized weights into the model first. This will help a lot later when we want to actually run the compressed models. Right now, when I load the model, I get a lot of warnings. 
1) The first warning shows that `weights_packed` and `weight_shape` were ignored in the state dict because the model doesn't have these attributes. This can be fixed by removing them from the unexpected keys but it would be great to use them to load the compressed model in `create_quantized_param` or making sure that the model have these field after using `apply_quantization_config` : 
```
Some weights of the model checkpoint at nm-testing/tinyllama-oneshot-w4a16-group128-v3 were not used when initializing LlamaForCausalLM: ['model.layers.0.mlp.down_proj.weight_packed', 'model.layers.0.mlp.down_proj.weight_shape',
```
2) The second warning complains about missing keys in the checkpoint: 

```
Some weights of LlamaForCausalLM were not initialized from the model checkpoint at nm-testing/tinyllama-oneshot-w4a16-group128-v3 and are newly initialized: ['model.layers.0.mlp.down_proj.weight', 'model.layers.0.mlp.down_proj.weight_zero_point'
```
This can be solved by fixing the missing keys using the `update_missing_keys` method. Also, note that here we are complaining that `weight_zero_point` even though the compressed model doesn't use zero-point at all.

- I think it will be better if we can decompress the model just based on the loaded quantized model only ! I don't think we want to go through the state dict again to decompress the model. 

These parts are a bit hard to deal with, so feel free to ask questions ! 

Interested in your thoughts about these two points @ArthurZucker

Reply:
Great points, thanks!  Will go through the missing/unexpected keys lists to try and get these resolved.

We may be able to do this by splitting up some of the logic in decompression to load based on the quantized model specifically however do not currently have this in place.
Note that in decompression we shouldn't be re-loading the entire state dict since we're going through safe-tensors and only calling/decompressing one layer at a time.  Let me know if this would work or if you have further questions.